src/topology/algebra/uniform_group.lean

```diff
@@ -0,0 +1,503 @@
+  begin
+    suffices : tendsto ((λp, -p) ∘ (λp:G×G, p.2 - p.1)) (comap (λp:G×G, p.2 - p.1) (nhds 0)) (nhds (-0)),
+    { simpa [(∘), tendsto_comap_iff] },
+    exact tendsto.comp tendsto_comap (tendsto_neg tendsto_id)
+  end,
+  comp                :=
+  begin
+    intros D H,
+    rw mem_lift'_sets,
+    { rcases H with ⟨U, U_nhds, U_sub⟩,
+      rcases exists_nhds_half U_nhds with ⟨V, ⟨V_nhds, V_sum⟩⟩,
+      existsi ((λp:G×G, p.2 - p.1) ⁻¹' V),
+      have H : (λp:G×G, p.2 - p.1) ⁻¹' V ∈ (comap (λp:G×G, p.2 - p.1) (nhds (0 : G))).sets,
+        by existsi [V, V_nhds] ; refl,
+      existsi H,
+      have comp_rel_sub : comp_rel ((λp:G×G, p.2 - p.1) ⁻¹' V) ((λp:G×G, p.2 - p.1) ⁻¹' V) ⊆ (λp:G×G, p.2 - p.1) ⁻¹' U,
+      begin
+        intros p p_comp_rel,
+        rcases p_comp_rel with ⟨z, ⟨Hz1, Hz2⟩⟩,
+        simpa using V_sum _ _ Hz1 Hz2
+      end,
+      exact set.subset.trans comp_rel_sub U_sub },
+    { exact monotone_comp_rel monotone_id monotone_id }
+  end,
+  is_open_uniformity  :=
+  begin
+    intro S,
+    let S' := λ x, {p : G × G | p.1 = x → p.2 ∈ S},
+    show is_open S ↔ ∀ (x : G), x ∈ S → S' x ∈ (comap (λp:G×G, p.2 - p.1) (nhds (0 : G))).sets,
+    rw [is_open_iff_mem_nhds],
+    refine forall_congr (assume a, forall_congr (assume ha, _)),
+    rw [← nhds_translation a, mem_comap_sets, mem_comap_sets],
+    refine exists_congr (assume t, exists_congr (assume ht, _)),
+    show (λ (y : G), y - a) ⁻¹' t ⊆ S ↔ (λ (p : G × G), p.snd - p.fst) ⁻¹' t ⊆ S' a,
+    split,
+    { rintros h ⟨x, y⟩ hx rfl, exact h hx },
+    { rintros h x hx, exact @h (a, x) hx rfl }
+  end }
+
+section
+local attribute [instance] topological_add_group.to_uniform_space
+
+lemma uniformity_eq_comap_nhds_zero' : uniformity = comap (λp:G×G, p.2 - p.1) (nhds (0 : G)) := rfl
+
+variable {G}
+lemma topological_add_group_is_uniform : uniform_add_group G :=
+have tendsto
+    ((λp:(G×G), p.1 - p.2) ∘ (λp:(G×G)×(G×G), (p.1.2 - p.1.1, p.2.2 - p.2.1)))
+    (comap (λp:(G×G)×(G×G), (p.1.2 - p.1.1, p.2.2 - p.2.1)) ((nhds 0).prod (nhds 0)))
+    (nhds (0 - 0)) :=
+  tendsto_comap.comp (tendsto_sub tendsto_fst tendsto_snd),
+begin
+  constructor,
+  rw [uniform_continuous, uniformity_prod_eq_prod, tendsto_map'_iff,
+    uniformity_eq_comap_nhds_zero' G, tendsto_comap_iff, prod_comap_comap_eq],
+  simpa [(∘)]
+end
+end
+
+lemma to_uniform_space_eq {α : Type*} [u : uniform_space α] [add_comm_group α] [uniform_add_group α]:
+  topological_add_group.to_uniform_space α = u :=
+begin
+  ext : 1,
+  show @uniformity α (topological_add_group.to_uniform_space α) = uniformity,
+  rw [uniformity_eq_comap_nhds_zero' α, uniformity_eq_comap_nhds_zero α]
+end
+
+end topological_add_comm_group
+
+
+
+
+namespace add_comm_group
+section Z_bilin
+
+variables {α : Type*} {β : Type*} {γ : Type*} {δ : Type*}
+variables [add_comm_group α] [add_comm_group β] [add_comm_group γ]
+
+/- TODO: when modules are changed to have more explicit base ring, then change replace `is_Z_bilin`
```

Comment:
Modules now have a more explicit base ring. I don't care if we fix this TODO during this refactor. We can also make a mental not to fix it after this refactor is merged.

Reply:
Yes, I'd like to keep this a pure reorganization PR.